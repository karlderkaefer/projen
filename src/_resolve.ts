import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
