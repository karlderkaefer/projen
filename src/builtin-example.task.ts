console.log("hello, I am a builtin task");
console.log("second line");
