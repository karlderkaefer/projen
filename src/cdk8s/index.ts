export * from "./auto-discover";
export * from "./cdk8s-app-ts";
export * from "./cdk8s-construct";
export * from "./integration-test";
