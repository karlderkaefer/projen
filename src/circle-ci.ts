import { Component } from './component';
import { Project } from './project';
import { YamlFile } from './yaml';

export interface CircleCiProps {
  readonly orbs?: Record<string, IOrb>;
  readonly description?: string;
  readonly jobs?: string[];
  readonly version: string;
  readonly workflows?: Record<string, IWorkflow>[];
}

export interface IOrb {
  readonly name: string;
  readonly version: string;
  toString(): string;
}

export class Orb implements IOrb {
  readonly name: string;
  readonly version: string;

  constructor(name: string, version: string) {
    this.name = name;
    this.version = version;
  }
  public toString(): string {
    return `${this.name}@${this.version}`;
  }
}

export interface IJob {
  readonly requires: string[];
  readonly context: string[];
  readonly filterBranchOnly: string[];
}

export interface IWorkflow {
  readonly jobs: Record<string, IJob[]>[];
}

function renderCircleCiFile(props?: CircleCiProps): object {
  const orbs : Record<string, string> = {};
  for (const [id, orb] of Object.entries(props?.orbs ?? {})) {
    orbs[id] = `${orb.toString()}`;
  }
  return {
    version: props?.version || '2.1',
    description: props?.description,
    orbs: orbs,
    jobs: props?.jobs,
    workflows: props?.workflows,
  };
}

export class CircleCi extends Component {
  constructor(project: Project, props?: CircleCiProps) {
    super(project);
    new YamlFile(project, 'circle.yml', {
      committed: true,
      readonly: true,
      obj: () => this._synthesizeCircleCi(props),
    });
  }
  _synthesizeCircleCi(props?: CircleCiProps): object {
    return renderCircleCiFile(props);
  }
}

