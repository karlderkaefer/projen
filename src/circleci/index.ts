export * from "./circleci";
export * from "./constant";
