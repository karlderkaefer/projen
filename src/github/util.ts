export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
