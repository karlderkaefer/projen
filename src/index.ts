export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
