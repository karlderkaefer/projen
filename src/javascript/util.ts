import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
