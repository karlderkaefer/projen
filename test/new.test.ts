// tests for `projen new`: we run `projen new` for each supported project type
// and compare against a golden snapshot.
import { execSync } from "child_process";
import { join } from "path";
import { pathExistsSync } from "fs-extra";
import * as inventory from "../src/inventory";
import { execCapture } from "../src/util";
import {
  directorySnapshot,
  execProjenCLI,
  synthSnapshot,
  synthSnapshotWithPost,
  TestProject,
  withProjectDir,
} from "./util";

for (const type of inventory.discover()) {
  test(`projen new ${type.pjid}`, () => {
    withProjectDir((projectdir) => {
      // execute `projen new PJID --no-synth` in the project directory
      execProjenCLI(projectdir, ["new", "--no-synth", type.pjid]);

      // compare generated .projenrc.js to the snapshot
      const actual = directorySnapshot(projectdir, {
        excludeGlobs: [".git/**"],
      });

      expect(actual).toMatchSnapshot();
    });
  });
}

test("post-synthesis option enabled", () => {
  const project = new TestProject();

  expect(synthSnapshotWithPost(project)[".postsynth"]).toContain("postsynth");
});

test("post-synthesis option disabled", () => {
  const project = new TestProject();

  expect(synthSnapshot(project)[".postsynth"]).toBeUndefined();
});

/**
 * commented out due to breaking changes in projen@0.37.0

test('projen new --from external', () => {
  withProjectDir(projectdir => {

    // execute `projen new --from cdk-appsync-project` in the project directory
    execProjenCLI(projectdir, ['new', '--from', 'cdk-appsync-project@1.1.3', '--no-post', MIN_NODE_VERSION_OPTION]);

    // patch the projen version in package.json to match the current version
    // otherwise, every bump would need to update these snapshots.
    sanitizeOutput(projectdir);

    // compare generated .projenrc.js to the snapshot
    const actual = directorySnapshot(projectdir, {
      excludeGlobs: [
        '.git/**',
        '.github/**',
        'node_modules/**',
        'yarn.lock',
      ],
    });

    expect(actual).toMatchSnapshot();
    expect(actual['schema.graphql']).toBeDefined();

  });
});

test('projen new --from external tarball', () => {
  withProjectDir(projectdir => {
    const shell = (command: string) => execSync(command, { cwd: projectdir });
    // downloads cdk-appsync-project-1.1.3.tgz
    shell('npm pack cdk-appsync-project@1.1.3');

    execProjenCLI(projectdir, ['new', '--from', './cdk-appsync-project-1.1.3.tgz', '--no-post', MIN_NODE_VERSION_OPTION]);

    // patch the projen version in package.json to match the current version
    // otherwise, every bump would need to update these snapshots.
    sanitizeOutput(projectdir);

    // compare generated .projenrc.js to the snapshot
    const actual = directorySnapshot(projectdir, {
      excludeGlobs: [
        '.git/**',
        '.github/**',
        'node_modules/**',
        'yarn.lock',
      ],
    });

    expect(actual['package.json']).toMatchSnapshot();
    expect(actual['.projenrc.js']).toMatchSnapshot();
    expect(actual['schema.graphql']).toBeDefined();
  });
});

test('projen new --from external dist tag', () => {
  withProjectDir(projectdir => {
    execProjenCLI(projectdir, ['new', '--from', 'cdk-appsync-project@latest', '--no-post', MIN_NODE_VERSION_OPTION]);

    // compare generated .projenrc.js to the snapshot
    const actual = directorySnapshot(projectdir, {
      excludeGlobs: [
        '.git/**',
        '.github/**',
        'node_modules/**',
        'yarn.lock',
      ],
    });

    // Not doing a snapshot test because @latest is used
    expect(actual['schema.graphql']).toBeDefined();
  });
});

test('options are not overwritten when creating from external project types', () => {
  withProjectDir(projectdir => {

    // execute `projen new --from cdk-appsync-project` in the project directory
    execProjenCLI(projectdir, ['new', '--from', 'cdk-appsync-project@1.1.3', '--no-synth', '--cdk-version', '1.63.0', MIN_NODE_VERSION_OPTION]);

    // compare generated .projenrc.js to the snapshot
    const actual = directorySnapshot(projectdir, {
      excludeGlobs: [
        '.git/**',
        '.github/**',
        'node_modules/**',
        'yarn.lock',
      ],
    });

    expect(actual['.projenrc.js']).toContain('cdkVersion: \'1.63.0\'');
  });
});

test('can choose from one of multiple external project types', () => {
  withProjectDir(projectdir => {

    // execute `projen new --from cdk-appsync-project` in the project directory
    execProjenCLI(projectdir, ['new', '--from', '@taimos/projen@0.0.126', 'taimos-ts-lib', '--no-post', MIN_NODE_VERSION_OPTION]);

    // patch the projen version in package.json to match the current version
    // otherwise, every bump would need to update these snapshots.
    sanitizeOutput(projectdir);

    // compare generated .projenrc.js to the snapshot
    const actual = directorySnapshot(projectdir, {
      excludeGlobs: [
        '.git/**',
        '.github/**',
        'node_modules/**',
        'yarn.lock',
      ],
    });

    expect(actual['.projenrc.js']).toContain('@taimos/projen@0.0.126');
  });
});

**/

test("projen new --no-comments", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, ["new", "node", "--no-comments", "--no-synth"]);

    const projenrc = directorySnapshot(projectdir)[".projenrc.js"];
    expect(projenrc).toBeDefined();
    expect(projenrc).not.toMatch("//");
  });
});

test("creating node project with enum-typed CLI arg", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, [
      "new",
      "node",
      "--package-manager",
      "npm",
      "--no-synth",
    ]);

    const projenrc = directorySnapshot(projectdir)[".projenrc.js"];
    expect(projenrc).toMatchSnapshot();
  });
});

test("creating python project with enum-typed CLI arg", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, [
      "new",
      "python",
      "--project-type",
      "lib",
      "--projenrc-python",
      "--no-synth",
    ]);

    const projenrc = directorySnapshot(projectdir)[".projenrc.py"];
    expect(projenrc).toMatchSnapshot();
  });
});

test("creating java project with enum-typed CLI arg", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, [
      "new",
      "java",
      "--project-type",
      "lib",
      "--projenrc-java",
      "--no-synth",
    ]);

    const projenrc =
      directorySnapshot(projectdir)["src/test/java/projenrc.java"];
    expect(projenrc).toMatchSnapshot();
  });
});

test("projenrc-json creates node-project", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, ["new", "node", "--projenrc-json", "--no-synth"]);

    const projenrc = directorySnapshot(projectdir)[".projenrc.json"];
    expect(projenrc).toMatchSnapshot();
  });
});

/**
 * commented out due to breaking changes in projen@0.37.0

test('projenrc-json creates external project type', () => {
  withProjectDir(projectdir => {
    execProjenCLI(projectdir, ['new', '--from', 'cdk-appsync-project@1.1.3', '--cdk-version', '1.63.0', '--projenrc-json', '--no-synth', MIN_NODE_VERSION_OPTION]);

    // exclude node_modules to work around bug where node_modules is generated AND one of the
    // dependencies includes a file with .json extension that isn't valid JSON
    const projenrc = directorySnapshot(projectdir, { excludeGlobs: ['node_modules/**'] })['.projenrc.json'];
    expect(projenrc).toMatchSnapshot();
  });
});

*/

test("projenrc-ts creates typescript projenrc", () => {
  withProjectDir((projectdir) => {
    execProjenCLI(projectdir, [
      "new",
      "typescript",
      "--projenrc-ts",
      "--no-synth",
    ]);
    const projenrc = directorySnapshot(projectdir)[".projenrc.ts"];
    expect(projenrc).toMatchSnapshot();
  });
});

test("projen new node --outdir path/to/mydir", () => {
  withProjectDir((projectdir) => {
    // GIVEN
    const shell = (command: string) => execSync(command, { cwd: projectdir });
    shell(`mkdir -p ${join("path", "to", "mydir")}`);

    // WHEN
    execProjenCLI(projectdir, ["new", "node", "--outdir", "path/to/mydir"]);

    // THEN
    const targetDirSnapshot = directorySnapshot(
      join(projectdir, "path", "to", "mydir"),
      { excludeGlobs: ["node_modules/**"] }
    );
    expect(targetDirSnapshot[".projenrc.js"]).toMatchSnapshot();
    expect(targetDirSnapshot["package.json"]).toBeDefined();
  });
});

describe("git", () => {
  test("--git (default) will initialize a git repo and create a commit", () => {
    withProjectDir((projectdir) => {
      execProjenCLI(projectdir, ["new", "project"]);
      expect(
        execCapture("git log", { cwd: projectdir })
          .toString("utf8")
          .includes("chore: project created with projen")
      ).toBeTruthy();
    });
  });

  test("--no-git will not create a git repo", () => {
    withProjectDir(
      (projectdir) => {
        execProjenCLI(projectdir, ["new", "project", "--no-git"]);
        expect(pathExistsSync(join(projectdir, ".git"))).toBeFalsy();
      },
      { git: false }
    );
  });
});
