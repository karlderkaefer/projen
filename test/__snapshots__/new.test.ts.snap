// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating java project with enum-typed CLI arg 1`] = `
"import io.github.cdklabs.projen.java.JavaProject;
import io.github.cdklabs.projen.java.JavaProjectOptions;
import io.github.cdklabs.projen.ProjectType;

public class projenrc {
    public static void main(String[] args) {
        JavaProject project = new JavaProject(JavaProjectOptions.builder()
            .artifactId(\\"my-app\\")
            .groupId(\\"org.acme\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .projectType(ProjectType.LIB)
            .projenrcJava(true)
            .build());
        project.synth();
    }
}"
`;

exports[`creating node project with enum-typed CLI arg 1`] = `
"const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  packageManager: javascript.NodePackageManager.NPM,

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();"
`;

exports[`creating python project with enum-typed CLI arg 1`] = `
"from projen.python import PythonProject
from projen import ProjectType

project = PythonProject(
    author_email=\\"my@user.email.com\\",
    author_name=\\"My User Name\\",
    module_name=\\"my_project\\",
    name=\\"my-project\\",
    version=\\"0.1.0\\",
    project_type=ProjectType.LIB,
    projenrc_python=True,
)

project.synth()"
`;

exports[`projen new awscdk-app-java 1`] = `
Object {
  "src/test/java/projenrc.java": "import io.github.cdklabs.projen.awscdk.AwsCdkJavaApp;
import io.github.cdklabs.projen.awscdk.AwsCdkJavaAppOptions;

public class projenrc {
    public static void main(String[] args) {
        AwsCdkJavaApp project = new AwsCdkJavaApp(AwsCdkJavaAppOptions.builder()
            .artifactId(\\"my-app\\")
            .cdkVersion(\\"2.1.0\\")
            .groupId(\\"org.acme\\")
            .mainClass(\\"org.acme.MyApp\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .build());
        project.synth();
    }
}",
}
`;

exports[`projen new awscdk-app-py 1`] = `
Object {
  ".projenrc.py": "from projen.awscdk import AwsCdkPythonApp

project = AwsCdkPythonApp(
    author_email=\\"my@user.email.com\\",
    author_name=\\"My User Name\\",
    cdk_version=\\"2.1.0\\",
    module_name=\\"my_project\\",
    name=\\"my-project\\",
    version=\\"0.1.0\\",
)

project.synth()",
}
`;

exports[`projen new awscdk-app-ts 1`] = `
Object {
  ".projenrc.js": "const { awscdk } = require(\\"projen\\");
const project = new awscdk.AwsCdkTypeScriptApp({
  cdkVersion: \\"2.1.0\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new awscdk-construct 1`] = `
Object {
  ".projenrc.js": "const { awscdk } = require(\\"projen\\");
const project = new awscdk.AwsCdkConstructLibrary({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdkVersion: \\"2.1.0\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new cdk8s-app-ts 1`] = `
Object {
  ".projenrc.js": "const { cdk8s } = require(\\"projen\\");
const project = new cdk8s.Cdk8sTypeScriptApp({
  cdk8sVersion: \\"1.0.0-beta.10\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new cdk8s-construct 1`] = `
Object {
  ".projenrc.js": "const { cdk8s } = require(\\"projen\\");
const project = new cdk8s.ConstructLibraryCdk8s({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdk8sVersion: \\"1.4.10\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new cdktf-construct 1`] = `
Object {
  ".projenrc.js": "const { cdktf } = require(\\"projen\\");
const project = new cdktf.ConstructLibraryCdktf({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdktfVersion: \\"^0.8.3\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new java 1`] = `
Object {
  "src/test/java/projenrc.java": "import io.github.cdklabs.projen.java.JavaProject;
import io.github.cdklabs.projen.java.JavaProjectOptions;

public class projenrc {
    public static void main(String[] args) {
        JavaProject project = new JavaProject(JavaProjectOptions.builder()
            .artifactId(\\"my-app\\")
            .groupId(\\"org.acme\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .build());
        project.synth();
    }
}",
}
`;

exports[`projen new jsii 1`] = `
Object {
  ".projenrc.js": "const { cdk } = require(\\"projen\\");
const project = new cdk.JsiiProject({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new nextjs 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.NextJsProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // tailwind: true,          /* Setup Tailwind CSS as a PostCSS plugin. */
});
project.synth();",
}
`;

exports[`projen new nextjs-ts 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.NextJsTypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // tailwind: true,          /* Setup Tailwind CSS as a PostCSS plugin. */
});
project.synth();",
}
`;

exports[`projen new node --outdir path/to/mydir 1`] = `
"const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();"
`;

exports[`projen new node 1`] = `
Object {
  ".projenrc.js": "const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new project 1`] = `Object {}`;

exports[`projen new python 1`] = `
Object {
  ".projenrc.py": "from projen.python import PythonProject

project = PythonProject(
    author_email=\\"my@user.email.com\\",
    author_name=\\"My User Name\\",
    module_name=\\"my_project\\",
    name=\\"my-project\\",
    version=\\"0.1.0\\",
)

project.synth()",
}
`;

exports[`projen new react 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.ReactProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new react-ts 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.ReactTypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new typescript 1`] = `
Object {
  ".projenrc.js": "const { typescript } = require(\\"projen\\");
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projen new typescript-app 1`] = `
Object {
  ".projenrc.js": "const { typescript } = require(\\"projen\\");
const project = new typescript.TypeScriptAppProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();",
}
`;

exports[`projenrc-json creates node-project 1`] = `
Object {
  "defaultReleaseBranch": "main",
  "name": "my-project",
  "projenrcJson": true,
  "type": "projen.javascript.NodeProject",
}
`;

exports[`projenrc-ts creates typescript projenrc 1`] = `
"import { typescript } from \\"projen\\";
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  projenrcTs: true,

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
});
project.synth();"
`;
